src/realm/deppart/byfield_gpu_impl.hpp

```diff
@@ -0,0 +1,261 @@
+                      RectFieldVolumeOp<N,T>,            // functor
+                      Rect<N,T>*     // underlying input iterator
+                    >;
+    VolIter d_volumes(d_valid_rects, RectFieldVolumeOp<N,T>());
+
+    // 3) get temporary storage size
+    void*   d_temp = nullptr;
+    size_t rect_temp_bytes = 0;
+    cub::DeviceScan::InclusiveSum(
+        /* d_temp_storage */  nullptr,
+        /* temp_bytes */      rect_temp_bytes,
+        /* d_in */            d_volumes,
+        /* d_out */           d_prefix_rects + 1,   // shift by one so prefix[1]..prefix[n]
+        /* num_items */       num_valid_rects, stream);
+
+    // 4) allocate and run
+    RegionInstance temp_instance = this->realm_malloc(rect_temp_bytes, my_mem);
+    d_temp = reinterpret_cast<void*>(AffineAccessor<char,1>(temp_instance, 0).base);
+    cub::DeviceScan::InclusiveSum(
+        /* d_temp_storage */  d_temp,
+        /* temp_bytes */      rect_temp_bytes,
+        /* d_in */            d_volumes,
+        /* d_out */           d_prefix_rects + 1,
+        /* num_items */       num_valid_rects, stream);
+
+
+    size_t total_pts;
+    CUDA_CHECK(cudaMemcpyAsync(&total_pts, &d_prefix_rects[num_valid_rects], sizeof(size_t), cudaMemcpyDeviceToHost, stream), stream);
+
+    CUDA_CHECK(cudaStreamSynchronize(stream), stream);
+
+    nvtx_range_pop();
+
+    nvtx_range_push("cuda", "build points");
+
+    RegionInstance points_instance = this->realm_malloc(total_pts * sizeof(PointDesc<N,T>), my_mem);
+    PointDesc<N,T>* d_points = reinterpret_cast<PointDesc<N,T>*>(AffineAccessor<char,1>(points_instance, 0).base);
+
+    FT* d_colors;
+    RegionInstance colors_instance;
+
+    if constexpr(std::is_same_v<FT,bool>) {
+      std::vector<uint8_t> flat_colors(colors.size());
+      for (size_t i = 0; i < colors.size(); i++) {
+        flat_colors[i] = colors[i] ? 1 : 0;
+      }
+      colors_instance = this->realm_malloc(total_pts * sizeof(PointDesc<N,T>), my_mem);
+      uint8_t* d_flat_colors = reinterpret_cast<uint8_t*>(AffineAccessor<char,1>(colors_instance, 0).base);
+      CUDA_CHECK(cudaMemcpyAsync(d_flat_colors, flat_colors.data(), colors.size() * sizeof(uint8_t), cudaMemcpyHostToDevice, stream), stream);
+      d_colors = reinterpret_cast<FT*>(d_flat_colors);
+    } else {
+      colors_instance = this->realm_malloc(colors.size() * sizeof(FT), my_mem);
+      d_colors = reinterpret_cast<FT*>(AffineAccessor<char,1>(colors_instance, 0).base);
+      CUDA_CHECK(cudaMemcpyAsync(d_colors, colors.data(), colors.size() * sizeof(FT), cudaMemcpyHostToDevice, stream), stream);
+    }
+
+    AffineAccessor<FT,N,T> h_accessors[field_data.size()];
+    for (size_t i = 0; i < field_data.size(); ++i) {
+      h_accessors[i] = AffineAccessor<FT,N,T>(field_data[i].inst, field_data[i].field_offset);
+    }
+
+    RegionInstance accessors_instance = this->realm_malloc(field_data.size() * sizeof(AffineAccessor<FT,N,T>), my_mem);
```

Comment:
this seems like an example of an instance that could probably live in zero copy memory instead of device memory to avoid a copy.

Reply:
Bumping this, could just be a zero copy vector of accessors.